proposals/explicit-backing-fields.md

```diff
@@ -0,0 +1,433 @@
+or primary property ([example](https://github.com/google/accompanist/blob/0d0198d0cab599295f7601cb386d5989d72cc8cd/pager/src/main/java/com/google/accompanist/pager/PagerState.kt#L145))
+is delegated.
+
+_Statistics_: backing property delegation found in 2693 files, primary property delegation in 1010 files.
+
+## Design
+
+### Explicit Backing Fields
+
+By analogy with the already existing term ["implicit backing fields"](https://kotlinlang.org/docs/properties.html#backing-properties), it is proposed to introduce a syntax
+for declaring "explicit backing fields" of properties.
+Syntax consists of a keyword `field`, optional type definition and initialization expression
+(more detailed [Grammar changes](#grammar-changes) are in the separate section).
+
+```kotlin
+private val _city = MutableLiveData<String>()
+val city: LiveData<String> get() = _city
+```
+
+Could be rewritten as follows:
+
+```kotlin
+val city: LiveData<String>
+    field = MutableLiveData()
+```
+
+### Visibility
+
+`private` is the default and the only allowed visibility for explicit backing fields
+(supporting other visibilities might be considered, see [Future enhancements](#future-enhancements)).
+
+Visibility of property must be more permissive than explicit backing field visibility.
+
+### Resolution
+
+Calls of properties with explicit backing field are resolved to
+- the backing field, if call is made from inside the class (within private scope),
+- getter, otherwise.
+
+```kotlin
+class SomeViewModel {
+    val city: LiveData<String>
+        field = MutableLiveData("")
+    
+    fun updateCity(newCity: String) {
+        city.value = newCity // visible as MutableLiveData, calling field
+    }
+}
+
+fun outside(vm: SomeViewModel) {
+    vm.city // visible as LiveData, calling getter
+} 
+```
+
+There is no possibility to call a getter inside the class
+(might be reconsidered in [Future enhancements](#future-enhancements)).
+
+### Accessors
+
+Property, which has explicit backing field, can also specify getter and/or setter.
+
+```kotlin
+val counter: Int
+    field = atomic(0)
+    get() = field.value
+```
+
+When explicit backing field with type `F` for a property with type `P` is declared,
+then compiler can derive getter and, for `var` properties, setter implementation automatically:
+
+* If `F` is a subtype of `P`, the compiler can provide a default getter.
+* If `P` is a subtype of `F`, the compiler can provide a default setter.
+
+If the compiler can not provide a getter, the user must declare it explicitly.
+The same applies to setters in case of `var` properties.
+
+```kotlin
+val city: String
+    field = MutableLiveData<String>()
+    get() = field.value ?: "-" // It is an error if getter is not explicitly specified here
+```
+
+### Explicit backing field in combination with property initializer
+
+It's allowed to have both property initializer and explicit backing field specified.
+
+```kotlin
+val city: StateFlow<String> = field.asStateFlow()
```

Comment:
What about cases this case:
```kotlin
val city: StateFlow<String> get() = field.asStateFlow()
    field = MutableStateFlow("")
```

Will it be supported? Now it's not really clear for me from the KEEP text

Reply:
Ah, indeed. Thanks